grep-0018-modernize-c++.md

```diff
@@ -0,0 +1,417 @@
+### More const
+
+Benefits:
+* more readable; even in a function reader can see that a name has a
+  final value
+* avoids accidents
+* possibly allows compiler optimizations
+
+Recommendation:
+* private members and other `_impl.xx` code should be free to change
+  at any time to add const
+* adding const to exposed interfaces can be done between minor
+  releases
+
+Risks: none.
+
+### More constexpr
+
+`constexpr` guarantees compile time evaluation.
+
+Benefits over const:
+* no risk of initialization order breakage
+* can be used in some places where const can't
+* enables compiler optimizations
+
+Recommendation: See "More const", above
+
+Risks: none.
+
+### Depointerization
+
+GNU Radio has a lot of member variables that are raw pointers. In many
+cases ([example][raw-pointer-to-obj]) they don't have to be pointers
+at all, which:
+* clarifies for the reader what the lifetime of the object is
+* removes pointer dereference at runtime
+* possibly enables compiler to make smarter optimizations
+* reduces need for manual memory management, and in many cases the
+  whole destructor
+
+Other raw pointers do need to remain dynamic since changing them
+requires instantiating a new object
+([example][raw-pointer-to-std]). They should be changed to `std` smart
+pointers, or even more preferably the objects should be movable and be
+changed that way. For objects where copies are cheap it's enough to be
+able to copy.
+
+Another class of raw pointers is arrays, which sometimes require
+volk-specific allocation. Now that `volk::vector` exists these can all
+be changed to `std::vector` or `volk::vector`.
+
+Recommendation:
+* Allow 3.9 release to break ABI by changing raw and Boost pointers to
+  `std` and non-pointers even in public interfaces.
+
+Risks: none.
+
+### Delete copy constructor / copy assignment
+
+Where classes are not copyable (e.g. contain a pointer), the copy
+constructor and copy assignment operator should be deleted.
+
+```c++
+  foo(const foo&) = delete;
+  foo& operator(const foo&) = delete;
+```
+
+This should be added even if the class has a `std::unique_ptr` member
+variable, since it may be changed to `std::shared_ptr` in the future.
+(TODO: opinions? how likely is that?)
+
+Recommendation:
+* Before minor version release, do this as much as possible
+* Where uncopyability is exposed in a released version and can be
+  fixed: make the object copyable. It's less efficient, but fixes
+  runtime behavior without breaking OOT builds.
+* Where uncopyability is inherent in a released version to the exposed
+  interface: break OOT builds by deleting copy constructor & copy
+  assignment. It's better to break at build time than runtime.
+
+Risks: none.
+
+### Boost smart pointers
+
+Boost smart pointers currently provide no extra benefit over C++11
+smart pointers. The only thing missing in C++11 (which arrived in
+C++14) is `std::make_unique()`. Until C++14 is required for GNU Radio
+that gap can be filled by `boost::make_unique()`.
```

Comment:
Or we implement our own `std::make_unique`.

Reply:
Added.